File: src/ui/brainbrawler.cpp
/* This file contains code for cracking  Bitcoin brainwallets with a dictionary attack
* This file contains the main entry point for the program
* 
* Author: Josh McIntyre
*/

#include "brainbrawler.h"

// This function is the main entry point for the program
int main(int argc, char* argv[])
{
	if (argc != 2)
	{
		output_help();
		exit(1);
	}

	std::string address = std::string(argv[1]);

	AttackResult result = attack(address);
	
	if (result.found == true)
	{
		std::cout  << "Password found: " << result.passphrase << std::endl;
	}
	else
	{
		std::cout  << "No password found for the specified brainwallet" << std::endl;
	}

	return 0;
}

void output_help()
{
	std::cout << "Usage: ./brainbrawler.exe <bitcoin address>";
}

File: src/py/brainbrawler.py
# This file contains code for cracking Bitcoin brainwallets with a dictionary attack
#
# Author: Josh McIntyre
#
import argparse

import bitaddrlib

WORDLIST_FILENAME = "wordlist_sample.txt"

class AttackResult:
    
    def __init__(self, found, passphrase, address):
        
        self.found = found
        self.passphrase = passphrase
        self.address = address

# Run a dictionary attack on possible brainwallets
def attack(address):

    bitaddr = bitaddrlib.BitAddr()

    with open(WORDLIST_FILENAME) as f:
        for word in f:
            word = word.strip()
            candidate_address = bitaddr.generate_brainwallet(word)
            if candidate_address == address:
                result = AttackResult(True, word, candidate_address)

                return result
        
        result = AttackResult(False, "", "")

        return result

# The main entry point for the program
def main():

    parser = argparse.ArgumentParser(description="Run a dictionary attack on a suspected Bitcoin brainwallet")
    parser.add_argument("address", type=str, help="The address to run the attack on")
    args = parser.parse_args()
    
    result = attack(args.address)
    
    if result.found:
        print(f"Password found: {result.passphrase}")
    else:
        print("No password found for the specified brainwallet")

if __name__ == "__main__":
	main()